design/sync-v2.md

```diff
@@ -0,0 +1,216 @@
+
+1. The state is set to attempted.
+1. The appropriate operation is performed on the remote service.
+    1. If the operation is successful, state is set to completed, and the replacement entity is written to the data table.
+    1. If the operation is failed, an error entry is set in the __errors table and the state is set to failed.
+
+Once the batch is complete:
+
+* Errors are compiled into an aggregate exception.
+* Any operation that is "completed" is removed from the operations queue.
+
+### Pull Process
+
+During the pull process, the sync process:
+
+1. Reads the delta token
+1. Does a "GET /tables/<table>?$filter=(updatedAt gt '<delta-token>')&__includedeleted=true - this results in a paged response.
+1. For each element in the paged response:
+    * Write the new or updated record to the data table (or delete the record).
+    * Store the delta token in the __config table.
+1. Repeat for additional pages of the paged response.
+
+## Problems with the current process
+
+1. It is highly reliable, but slow.
+1. The current SQLite implementation does not allow for additional property types (like spatial data, DateOnly, or TimeOnly)
+1. Requires full support of OData to SQL to work.
+1. Requires a specific form of the DTO to properly support incremental sync, conflict resolution, and optimistic concurrency.
+
+## Proposal
+
+Let's look at the remote service end and the local offline end separately.
+
+### Remote Service
+
+By using the OData service endpoint, we allow filtering (thus allowing just a subset of records to be synchronized) and projection (thus allowing only certain properties to be synchronized).  Ideally, we could use any DTO for synchronization.  For example, this is not available today:
+
+```csharp
+class MyDto 
+{
+    [Key]
+    public int localId { get; set; }
+}
+```
+
+* The localId is not a string (and not globally unique) and not called `Id`.
+* There is no `UpdatedAt` or `Version` strings.
+
+I'd propose the following:
+
+* The ID should be identified by a `[Key]` attribute and can be any object.  It is up to the developer to ensure that the ID is globally unique within the system, including offline.
+* The last updated date should be identified by a new attribute `[UpdatedAt]` and must be an auto-incrementing comparable.  For example, we can use a `long` or a `DateTimeOffset`.  The only requirement is that it must be every increasing so we can use a `>` comparison.
+* The version must either be specified by a `[Version]` filter (and can either be a string or byte array), or the DTO must implement `IDatasyncVersion` which has a `.GetVersion()` (returns a string) to generate the version.  This allows the developer to decide how to compute a version.
+* A `Deleted` flag may be provided by using `[Deleted]` on a field.
+
+For a `TodoItem`, for example:
+
+```csharp
+class TodoItemDTO : IDatasyncVersion
+{
+    [Key]
+    public long Id { get; set; }
```

Comment:
Opening the client to other Id field types (`int`/`long`) could open the way to indexes in SQLite and other dbs? 
I guess, `IDatasyncVersion.GetId()` is needed to support the developer in the painful process to ensure a globally unique id?